File: rust/hron/src/lexer.rs
use crate::error::{ScheduleError, Span};

/// Token produced by the lexer.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TokenKind {
    // Keywords
    Every,
    On,
    At,
    From,
    To,
    In,
    Of,
    The,
    Last,
    Except,
    Until,
    Starting,
    During,
    Year,

    // Day keywords
    Day,
    Weekday,
    Weekend,
    Weeks,
    Month,

    // Day names
    DayName(String), // lowercase full name: "monday", "tuesday", ...

    // Month names
    MonthName(String), // lowercase short: "jan", "feb", ...

    // Ordinals
    Ordinal(String), // "first", "second", "third", "fourth", "fifth"

    // Interval units
    IntervalUnit(String), // "min", "mins", "minute", "minutes", "hour", "hours", "hr", "hrs"

    // Literals
    Number(u32),
    OrdinalNumber(u32), // 1st, 2nd, 3rd, 15th — the number part
    Time(u8, u8),       // HH:MM
    IsoDate(String),    // 2026-03-15

    // Punctuation
    Comma,

    // Timezone (IANA string)
    Timezone(String),
}

pub struct Lexer<'a> {
    input: &'a str,
    bytes: &'a [u8],
    pos: usize,
    /// Set after we emit an `In` token so we know to parse a timezone next.
    after_in: bool,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            input,
            bytes: input.as_bytes(),
            pos: 0,
            after_in: false,
        }
    }

    pub fn tokenize(&mut self) -> Result<Vec<Token>, ScheduleError> {
        let mut tokens = Vec::new();
        loop {
            self.skip_whitespace();
            if self.pos >= self.bytes.len() {
                break;
            }

            // After `in` keyword, consume the rest as a timezone string
            if self.after_in {
                self.after_in = false;
                let tok = self.lex_timezone()?;
                tokens.push(tok);
                continue;
            }

            let start = self.pos;
            let ch = self.bytes[self.pos];

            if ch == b',' {
                self.pos += 1;
                tokens.push(Token {
                    kind: TokenKind::Comma,
                    span: Span::new(start, self.pos),
                });
                continue;
            }

            // Try time literal: HH:MM (but not ISO date YYYY-MM-DD)
            if ch.is_ascii_digit() {
                let tok = self.lex_number_or_time_or_date()?;
                tokens.push(tok);
                continue;
            }

            // Word
            if ch.is_ascii_alphabetic() {
                let tok = self.lex_word()?;
                tokens.push(tok);
                continue;
            }

            return Err(ScheduleError::lex(
                format!("unexpected character '{}'", ch as char),
                Span::new(start, start + 1),
                self.input,
            ));
        }
        Ok(tokens)
    }

    fn skip_whitespace(&mut self) {
        while self.pos < self.bytes.len() && self.bytes[self.pos].is_ascii_whitespace() {
            self.pos += 1;
        }
    }

    fn lex_timezone(&mut self) -> Result<Token, ScheduleError> {
        self.skip_whitespace();
        let start = self.pos;
        // Consume everything remaining as the timezone
        while self.pos < self.bytes.len() && !self.bytes[self.pos].is_ascii_whitespace() {
            self.pos += 1;
        }
        // IANA timezones are single tokens like "America/Vancouver" or "UTC", no spaces.
        let tz = &self.input[start..self.pos];
        if tz.is_empty() {
            return Err(ScheduleError::lex(
                "expected timezone after 'in'",
                Span::new(start, start + 1),
                self.input,
            ));
        }
        Ok(Token {
            kind: TokenKind::Timezone(tz.to_string()),
            span: Span::new(start, self.pos),
        })
    }

    fn lex_number_or_time_or_date(&mut self) -> Result<Token, ScheduleError> {
        let start = self.pos;
        // Read digits
        let num_start = self.pos;
        while self.pos < self.bytes.len() && self.bytes[self.pos].is_ascii_digit() {
            self.pos += 1;
        }
        let digits = &self.input[num_start..self.pos];

        // Check for ISO date: YYYY-MM-DD
        if digits.len() == 4
            && self.pos < self.bytes.len()
            && self.bytes[self.pos] == b'-'
            && self.pos + 3 <= self.bytes.len()
        {
            // Peek ahead to see if this is YYYY-MM-DD
            let remaining = &self.input[self.pos..];
            if remaining.len() >= 6 {
                let maybe_date = &self.input[start..];
                if maybe_date.len() >= 10
                    && maybe_date.as_bytes()[4] == b'-'
                    && maybe_date.as_bytes()[5].is_ascii_digit()
                    && maybe_date.as_bytes()[6].is_ascii_digit()
                    && maybe_date.as_bytes()[7] == b'-'
                    && maybe_date.as_bytes()[8].is_ascii_digit()
                    && maybe_date.as_bytes()[9].is_ascii_digit()
                {
                    self.pos = start + 10;
                    return Ok(Token {
                        kind: TokenKind::IsoDate(self.input[start..self.pos].to_string()),
                        span: Span::new(start, self.pos),
                    });
                }
            }
        }

        // Check for time: HH:MM
        if (digits.len() == 1 || digits.len() == 2)
            && self.pos < self.bytes.len()
            && self.bytes[self.pos] == b':'
        {
            self.pos += 1; // skip ':'
            let min_start = self.pos;
            while self.pos < self.bytes.len() && self.bytes[self.pos].is_ascii_digit() {
                self.pos += 1;
            }
            let min_digits = &self.input[min_start..self.pos];
            if min_digits.len() == 2 {
                let hour: u8 = digits.parse().map_err(|_| {
                    ScheduleError::lex("invalid hour", Span::new(start, self.pos), self.input)
                })?;
                let minute: u8 = min_digits.parse().map_err(|_| {
                    ScheduleError::lex("invalid minute", Span::new(start, self.pos), self.input)
                })?;
                if hour > 23 || minute > 59 {
                    return Err(ScheduleError::lex(
                        "invalid time",
                        Span::new(start, self.pos),
                        self.input,
                    ));
                }
                return Ok(Token {
                    kind: TokenKind::Time(hour, minute),
                    span: Span::new(start, self.pos),
                });
            }
        }

        let num: u32 = digits.parse().map_err(|_| {
            ScheduleError::lex("invalid number", Span::new(start, self.pos), self.input)
        })?;

        // Check for ordinal suffix: st, nd, rd, th
        if self.pos + 1 < self.bytes.len() {
            let suffix = &self.input[self.pos..self.pos + 2].to_lowercase();
            if matches!(suffix.as_str(), "st" | "nd" | "rd" | "th") {
                self.pos += 2;
                return Ok(Token {
                    kind: TokenKind::OrdinalNumber(num),
                    span: Span::new(start, self.pos),
                });
            }
        }

        Ok(Token {
            kind: TokenKind::Number(num),
            span: Span::new(start, self.pos),
        })
    }

    fn lex_word(&mut self) -> Result<Token, ScheduleError> {
        let start = self.pos;
        while self.pos < self.bytes.len()
            && (self.bytes[self.pos].is_ascii_alphanumeric() || self.bytes[self.pos] == b'_')
        {
            self.pos += 1;
        }
        let word = self.input[start..self.pos].to_lowercase();

        let kind = match word.as_str() {
            "every" => TokenKind::Every,
            "on" => TokenKind::On,
            "at" => TokenKind::At,
            "from" => TokenKind::From,
            "to" => TokenKind::To,
            "in" => {
                self.after_in = true;
                TokenKind::In
            }
            "of" => TokenKind::Of,
            "the" => TokenKind::The,
            "last" => TokenKind::Last,
            "except" => TokenKind::Except,
            "until" => TokenKind::Until,
            "starting" => TokenKind::Starting,
            "during" => TokenKind::During,
            "year" => TokenKind::Year,

            "day" => TokenKind::Day,
            "weekday" | "weekdays" => TokenKind::Weekday,
            "weekend" | "weekends" => TokenKind::Weekend,
            "weeks" | "week" => TokenKind::Weeks,
            "month" => TokenKind::Month,

            "monday" | "mon" => TokenKind::DayName("monday".into()),
            "tuesday" | "tue" => TokenKind::DayName("tuesday".into()),
            "wednesday" | "wed" => TokenKind::DayName("wednesday".into()),
            "thursday" | "thu" => TokenKind::DayName("thursday".into()),
            "friday" | "fri" => TokenKind::DayName("friday".into()),
            "saturday" | "sat" => TokenKind::DayName("saturday".into()),
            "sunday" | "sun" => TokenKind::DayName("sunday".into()),

            "january" | "jan" => TokenKind::MonthName("jan".into()),
            "february" | "feb" => TokenKind::MonthName("feb".into()),
            "march" | "mar" => TokenKind::MonthName("mar".into()),
            "april" | "apr" => TokenKind::MonthName("apr".into()),
            "may" => TokenKind::MonthName("may".into()),
            "june" | "jun" => TokenKind::MonthName("jun".into()),
            "july" | "jul" => TokenKind::MonthName("jul".into()),
            "august" | "aug" => TokenKind::MonthName("aug".into()),
            "september" | "sep" => TokenKind::MonthName("sep".into()),
            "october" | "oct" => TokenKind::MonthName("oct".into()),
            "november" | "nov" => TokenKind::MonthName("nov".into()),
            "december" | "dec" => TokenKind::MonthName("dec".into()),

            "first" => TokenKind::Ordinal("first".into()),
            "second" => TokenKind::Ordinal("second".into()),
            "third" => TokenKind::Ordinal("third".into()),
            "fourth" => TokenKind::Ordinal("fourth".into()),
            "fifth" => TokenKind::Ordinal("fifth".into()),

            "min" | "mins" | "minute" | "minutes" => TokenKind::IntervalUnit("min".into()),
            "hour" | "hours" | "hr" | "hrs" => TokenKind::IntervalUnit("hours".into()),

            _ => {
                return Err(ScheduleError::lex(
                    format!("unknown keyword '{word}'"),
                    Span::new(start, self.pos),
                    self.input,
                ));
            }
        };

        Ok(Token {
            kind,
            span: Span::new(start, self.pos),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_day_repeat() {
        let mut lexer = Lexer::new("every day at 09:00");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens.len(), 4);
        assert_eq!(tokens[0].kind, TokenKind::Every);
        assert_eq!(tokens[1].kind, TokenKind::Day);
        assert_eq!(tokens[2].kind, TokenKind::At);
        assert_eq!(tokens[3].kind, TokenKind::Time(9, 0));
    }

    #[test]
    fn test_iso_date() {
        let mut lexer = Lexer::new("on 2026-03-15 at 14:30");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[1].kind, TokenKind::IsoDate("2026-03-15".into()));
        assert_eq!(tokens[3].kind, TokenKind::Time(14, 30));
    }

    #[test]
    fn test_ordinal_number() {
        let mut lexer = Lexer::new("every month on the 1st at 09:00");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[3].kind, TokenKind::The);
        assert_eq!(tokens[4].kind, TokenKind::OrdinalNumber(1));
    }

    #[test]
    fn test_timezone() {
        let mut lexer = Lexer::new("every day at 09:00 in America/Vancouver");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(
            tokens.last().unwrap().kind,
            TokenKind::Timezone("America/Vancouver".into())
        );
    }

    #[test]
    fn test_interval() {
        let mut lexer = Lexer::new("every 30 min from 09:00 to 17:00");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[1].kind, TokenKind::Number(30));
        assert_eq!(tokens[2].kind, TokenKind::IntervalUnit("min".into()));
    }

    #[test]
    fn test_except_token() {
        let mut lexer = Lexer::new("every weekday at 09:00 except dec 25");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[4].kind, TokenKind::Except);
        assert_eq!(tokens[5].kind, TokenKind::MonthName("dec".into()));
    }

    #[test]
    fn test_until_token() {
        let mut lexer = Lexer::new("every day at 09:00 until 2026-12-31");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[4].kind, TokenKind::Until);
        assert_eq!(tokens[5].kind, TokenKind::IsoDate("2026-12-31".into()));
    }

    #[test]
    fn test_starting_token() {
        let mut lexer = Lexer::new("every 2 weeks on monday at 09:00 starting 2026-01-05");
        let tokens = lexer.tokenize().unwrap();
        assert!(tokens.iter().any(|t| t.kind == TokenKind::Starting));
    }

    #[test]
    fn test_year_token() {
        let mut lexer = Lexer::new("every year on dec 25 at 00:00");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[1].kind, TokenKind::Year);
    }
}

File: go/lexer.go
package hron

import (
	"strconv"
	"strings"
)

// TokenKind represents the type of token.
type TokenKind int

const (
	TokenEvery TokenKind = iota
	TokenOn
	TokenAt
	TokenFrom
	TokenTo
	TokenIn
	TokenOf
	TokenThe
	TokenLast
	TokenExcept
	TokenUntil
	TokenStarting
	TokenDuring
	TokenYear
	TokenDay
	TokenWeekday
	TokenWeekend
	TokenWeeks
	TokenMonth
	TokenDayName
	TokenMonthName
	TokenOrdinal
	TokenIntervalUnit
	TokenNumber
	TokenOrdinalNumber
	TokenTime
	TokenISODate
	TokenComma
	TokenTimezone
	TokenNearest
	TokenNext
	TokenPrevious
)

// Token represents a lexed token.
type Token struct {
	Kind TokenKind
	Span Span

	// Value fields (only one is set based on Kind)
	DayNameVal   Weekday
	MonthNameVal MonthName
	OrdinalVal   OrdinalPosition
	UnitVal      IntervalUnit
	NumberVal    int
	TimeHour     int
	TimeMinute   int
	ISODateVal   string
	TimezoneVal  string
}

// lexer is the internal lexer state.
type lexer struct {
	input   string
	pos     int
	afterIn bool
}

// Tokenize tokenizes the input string into a list of tokens.
func Tokenize(input string) ([]Token, error) {
	l := &lexer{input: input}
	return l.tokenize()
}

func (l *lexer) tokenize() ([]Token, error) {
	var tokens []Token
	for {
		l.skipWhitespace()
		if l.pos >= len(l.input) {
			break
		}

		if l.afterIn {
			l.afterIn = false
			tok, err := l.lexTimezone()
			if err != nil {
				return nil, err
			}
			tokens = append(tokens, tok)
			continue
		}

		start := l.pos
		ch := l.input[l.pos]

		if ch == ',' {
			l.pos++
			tokens = append(tokens, Token{Kind: TokenComma, Span: Span{start, l.pos}})
			continue
		}

		if isDigit(ch) {
			tok, err := l.lexNumberOrTimeOrDate()
			if err != nil {
				return nil, err
			}
			tokens = append(tokens, tok)
			continue
		}

		if isAlpha(ch) {
			tok, err := l.lexWord()
			if err != nil {
				return nil, err
			}
			tokens = append(tokens, tok)
			continue
		}

		return nil, LexError("unexpected character '"+string(ch)+"'", Span{start, start + 1}, l.input)
	}

	return tokens, nil
}

func (l *lexer) skipWhitespace() {
	for l.pos < len(l.input) && isWhitespace(l.input[l.pos]) {
		l.pos++
	}
}

func (l *lexer) lexTimezone() (Token, error) {
	l.skipWhitespace()
	start := l.pos
	for l.pos < len(l.input) && !isWhitespace(l.input[l.pos]) {
		l.pos++
	}
	tz := l.input[start:l.pos]
	if len(tz) == 0 {
		return Token{}, LexError("expected timezone after 'in'", Span{start, start + 1}, l.input)
	}
	return Token{Kind: TokenTimezone, Span: Span{start, l.pos}, TimezoneVal: tz}, nil
}

func (l *lexer) lexNumberOrTimeOrDate() (Token, error) {
	start := l.pos

	// Read digits
	numStart := l.pos
	for l.pos < len(l.input) && isDigit(l.input[l.pos]) {
		l.pos++
	}
	digits := l.input[numStart:l.pos]

	// Check for ISO date: YYYY-MM-DD
	if len(digits) == 4 && l.pos < len(l.input) && l.input[l.pos] == '-' {
		remaining := l.input[start:]
		if len(remaining) >= 10 &&
			remaining[4] == '-' &&
			isDigit(remaining[5]) &&
			isDigit(remaining[6]) &&
			remaining[7] == '-' &&
			isDigit(remaining[8]) &&
			isDigit(remaining[9]) {
			l.pos = start + 10
			return Token{Kind: TokenISODate, Span: Span{start, l.pos}, ISODateVal: l.input[start:l.pos]}, nil
		}
	}

	// Check for time: HH:MM
	if (len(digits) == 1 || len(digits) == 2) && l.pos < len(l.input) && l.input[l.pos] == ':' {
		l.pos++ // skip ':'
		minStart := l.pos
		for l.pos < len(l.input) && isDigit(l.input[l.pos]) {
			l.pos++
		}
		minDigits := l.input[minStart:l.pos]
		if len(minDigits) == 2 {
			hour, err := strconv.Atoi(digits)
			if err != nil {
				return Token{}, LexError("invalid time hour", Span{start, l.pos}, l.input)
			}
			minute, err := strconv.Atoi(minDigits)
			if err != nil {
				return Token{}, LexError("invalid time minute", Span{start, l.pos}, l.input)
			}
			if hour > 23 || minute > 59 {
				return Token{}, LexError("invalid time", Span{start, l.pos}, l.input)
			}
			return Token{Kind: TokenTime, Span: Span{start, l.pos}, TimeHour: hour, TimeMinute: minute}, nil
		}
	}

	num, err := strconv.Atoi(digits)
	if err != nil {
		return Token{}, LexError("invalid number", Span{start, l.pos}, l.input)
	}

	// Check for ordinal suffix: st, nd, rd, th
	if l.pos+1 < len(l.input) {
		suffix := strings.ToLower(l.input[l.pos : l.pos+2])
		if suffix == "st" || suffix == "nd" || suffix == "rd" || suffix == "th" {
			l.pos += 2
			return Token{Kind: TokenOrdinalNumber, Span: Span{start, l.pos}, NumberVal: num}, nil
		}
	}

	return Token{Kind: TokenNumber, Span: Span{start, l.pos}, NumberVal: num}, nil
}

func (l *lexer) lexWord() (Token, error) {
	start := l.pos
	for l.pos < len(l.input) && (isAlphanumeric(l.input[l.pos]) || l.input[l.pos] == '_') {
		l.pos++
	}
	word := strings.ToLower(l.input[start:l.pos])
	span := Span{start, l.pos}

	// Check keyword map
	tok, ok := keywordMap[word]
	if !ok {
		return Token{}, LexError("unknown keyword '"+word+"'", span, l.input)
	}

	tok.Span = span

	if tok.Kind == TokenIn {
		l.afterIn = true
	}

	return tok, nil
}

// keywordMap maps lowercase keywords to tokens.
var keywordMap = map[string]Token{
	"every":    {Kind: TokenEvery},
	"on":       {Kind: TokenOn},
	"at":       {Kind: TokenAt},
	"from":     {Kind: TokenFrom},
	"to":       {Kind: TokenTo},
	"in":       {Kind: TokenIn},
	"of":       {Kind: TokenOf},
	"the":      {Kind: TokenThe},
	"last":     {Kind: TokenLast},
	"except":   {Kind: TokenExcept},
	"until":    {Kind: TokenUntil},
	"starting": {Kind: TokenStarting},
	"during":   {Kind: TokenDuring},
	"year":     {Kind: TokenYear},
	"years":    {Kind: TokenYear},
	"day":      {Kind: TokenDay},
	"days":     {Kind: TokenDay},
	"weekday":  {Kind: TokenWeekday},
	"weekdays": {Kind: TokenWeekday},
	"weekend":  {Kind: TokenWeekend},
	"weekends": {Kind: TokenWeekend},
	"weeks":    {Kind: TokenWeeks},
	"week":     {Kind: TokenWeeks},
	"month":    {Kind: TokenMonth},
	"months":   {Kind: TokenMonth},
	// Day names
	"monday":    {Kind: TokenDayName, DayNameVal: Monday},
	"mon":       {Kind: TokenDayName, DayNameVal: Monday},
	"tuesday":   {Kind: TokenDayName, DayNameVal: Tuesday},
	"tue":       {Kind: TokenDayName, DayNameVal: Tuesday},
	"wednesday": {Kind: TokenDayName, DayNameVal: Wednesday},
	"wed":       {Kind: TokenDayName, DayNameVal: Wednesday},
	"thursday":  {Kind: TokenDayName, DayNameVal: Thursday},
	"thu":       {Kind: TokenDayName, DayNameVal: Thursday},
	"friday":    {Kind: TokenDayName, DayNameVal: Friday},
	"fri":       {Kind: TokenDayName, DayNameVal: Friday},
	"saturday":  {Kind: TokenDayName, DayNameVal: Saturday},
	"sat":       {Kind: TokenDayName, DayNameVal: Saturday},
	"sunday":    {Kind: TokenDayName, DayNameVal: Sunday},
	"sun":       {Kind: TokenDayName, DayNameVal: Sunday},
	// Month names
	"january":   {Kind: TokenMonthName, MonthNameVal: Jan},
	"jan":       {Kind: TokenMonthName, MonthNameVal: Jan},
	"february":  {Kind: TokenMonthName, MonthNameVal: Feb},
	"feb":       {Kind: TokenMonthName, MonthNameVal: Feb},
	"march":     {Kind: TokenMonthName, MonthNameVal: Mar},
	"mar":       {Kind: TokenMonthName, MonthNameVal: Mar},
	"april":     {Kind: TokenMonthName, MonthNameVal: Apr},
	"apr":       {Kind: TokenMonthName, MonthNameVal: Apr},
	"may":       {Kind: TokenMonthName, MonthNameVal: May},
	"june":      {Kind: TokenMonthName, MonthNameVal: Jun},
	"jun":       {Kind: TokenMonthName, MonthNameVal: Jun},
	"july":      {Kind: TokenMonthName, MonthNameVal: Jul},
	"jul":       {Kind: TokenMonthName, MonthNameVal: Jul},
	"august":    {Kind: TokenMonthName, MonthNameVal: Aug},
	"aug":       {Kind: TokenMonthName, MonthNameVal: Aug},
	"september": {Kind: TokenMonthName, MonthNameVal: Sep},
	"sep":       {Kind: TokenMonthName, MonthNameVal: Sep},
	"october":   {Kind: TokenMonthName, MonthNameVal: Oct},
	"oct":       {Kind: TokenMonthName, MonthNameVal: Oct},
	"november":  {Kind: TokenMonthName, MonthNameVal: Nov},
	"nov":       {Kind: TokenMonthName, MonthNameVal: Nov},
	"december":  {Kind: TokenMonthName, MonthNameVal: Dec},
	"dec":       {Kind: TokenMonthName, MonthNameVal: Dec},
	// Ordinals
	"first":  {Kind: TokenOrdinal, OrdinalVal: First},
	"second": {Kind: TokenOrdinal, OrdinalVal: Second},
	"third":  {Kind: TokenOrdinal, OrdinalVal: Third},
	"fourth": {Kind: TokenOrdinal, OrdinalVal: Fourth},
	"fifth":  {Kind: TokenOrdinal, OrdinalVal: Fifth},
	// Nearest weekday keywords
	"nearest":  {Kind: TokenNearest},
	"next":     {Kind: TokenNext},
	"previous": {Kind: TokenPrevious},
	// Interval units
	"min":     {Kind: TokenIntervalUnit, UnitVal: IntervalMin},
	"mins":    {Kind: TokenIntervalUnit, UnitVal: IntervalMin},
	"minute":  {Kind: TokenIntervalUnit, UnitVal: IntervalMin},
	"minutes": {Kind: TokenIntervalUnit, UnitVal: IntervalMin},
	"hour":    {Kind: TokenIntervalUnit, UnitVal: IntervalHours},
	"hours":   {Kind: TokenIntervalUnit, UnitVal: IntervalHours},
	"hr":      {Kind: TokenIntervalUnit, UnitVal: IntervalHours},
	"hrs":     {Kind: TokenIntervalUnit, UnitVal: IntervalHours},
}

// Helper functions

func isDigit(b byte) bool {
	return b >= '0' && b <= '9'
}

func isAlpha(b byte) bool {
	return (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z')
}

func isAlphanumeric(b byte) bool {
	return isAlpha(b) || isDigit(b)
}

func isWhitespace(b byte) bool {
	return b == ' ' || b == '\t' || b == '\n' || b == '\r'
}

